### ISSUE TITLE
Expressing the idea that a function successfully returned nothing is difficult 

### ISSUE THREAD

**Issue by codesections** [Raku/problem-solving#275]
Most functions in Raku return values, but a few don't.  In for those functions, we have [`Nil`](https://docs.raku.org/type/Nil) and its subtype, [`Failure`](https://docs.raku.org/type/Failure).  Here are a few examples of functions that return something that `.isa(Nil)`:

```raku
#`(1) sleep(0).isa(Nil);
#`(2) ‘foo’.index('b').isa(Nil);
#`(3) ‘’.unival.isa(Nil);
#`(4) sub f {}; &f.WHY.isa(Nil);
#`(5) (+'tree').isa(Nil);
#`(6) dd.isa(Nil);
#`(7) EVAL('42', :check).isa(Nil);
```

In each case, the function does not return a value, and so returning the special value `Nil` to signal an absence is 100% appropriate.  However, even though `Nil` always means "there wasn't a value", these functions really fall into three different categories: 
* `Nil` signals an error – we didn't get a value, but should have (ex. 5) 
* `Nil` signals successful completion – just like on the command line, no news is good news (ex.s 1, 6, and 7)
* `Nil` doesn't signals success or failure, it just means "no value found". (ex.s 2, 3, and 4)

I could have given more error examples, but I didn't bother because Raku already has `Failure` to distinguish between that category from the others.  However, we _don't_ currently have a good way to distinguish between the other two categories of `Nil`.  I believe that having a way to do so would help make our type system both slightly easier to understand for new users and slightly more powerful for more experienced users.

**Comment by codesections** [comment]
My proposal for addressing this issue is very simple: Add a new `Ok` or `Success` type that is the exact inverse of a `Failure` – that is, it is a subtype of `Nil` but signals successful completion of an operation.  A minimal implementation of this type could be as simple as:

```raku
class Ok is Nil { 
    method Bool(--> True) {} 
    method gist() { self.^name } 
}
```

## Advantages 

Adding this class (and, where appropriate, changing return values) would have two advantages:
### More readable code, (especially in combination with `try` or other `Failures`)

For example, both of these snippets would become possible:

```raku
#`( perform actions) if try EVAL($foo, :check); 
```
and
```raku
my %h is BagHash;
(try %h.add: +$str ) ?? #`(handle success) !! #`(handle failure);
```

### More self-documenting method/subroutine signatures

Right now, many functions have signatures that look a lot like

```raku
sub sleep($seconds = Inf --> Nil)
```

That's _fairly_ obvious, once you think a bit about what `Nil` must mean in that context.  But (imo), the following would be much clearer, especially to new users:

```raku
sub sleep($seconds = Inf --> Ok)
```
This difference may seem – and, indeed, be – pretty minor in the context of simple subs like `sleep`.  However, in more complex function signatures (which will be increasingly prevalent as Raku begins to be used in production for larger and larger projects), the added clarity of returning `Ok` instead of `Nil` seems like it would be even more helpful.

## Potential downsides

The only downside I can think of is that this change adds one more core type, and thus slightly increases the complexity of the Raku type system, which risks making the language more difficult to learn.  (Thanks to @lizmat for pointing this issue out on IRC).  However, I believe that two factors mitigate against this consideration.  First, Raku currently has a very rich type system with [over 300 built-in types](https://docs.raku.org/type.html); adding one more with a self-explanatory name does not seem likely to increase the complexity significantly.

Second and more importantly, splitting out a separate `Ok` class could _remove_ a significant source of confusion for new Rakoons.  I have observed that many new Rakoons think that `Nil` always indicates a failure.  Indeed, the docs (correctly!) [note that](https://docs.raku.org/type/Nil) "[a `Nil`] may also be used as a cheaper and less explosive alternative to a `Failure`"  and (correct!) [statements like](https://stackoverflow.com/questions/55072228#55079804) "Nil is really a type of generic soft failure" are pretty common.  While these statements are, as noted, correct, they could confuse users into thinking that a `Nil` always represents a `Failure` or at least an unexpectedly missing value.  Adding an `Ok` (or similar) type could help avoid this confusion and, hopefully, simplify learning Raku.

## Past discussions

I previously discussed this suggestion on the [#raku-dev](https://colabti.org/irclogger/irclogger_log/raku-dev?date=2021-03-07#l102) channel.  I tried to summarize the points raised in that discussion, but the logs are linked above if you want the full discussion.

**Comment by vrurg** [comment]
This needs some careful thinking. So far, not sure if I really like it, even though I well understand the reasoning behind it.

The first thing which would confuse me is a type object being `True`. Correct me if I'm wrong, but so far we have no class being `True` without instantiation. That's definitely will be a confusion for a newby.

Aside of that, semantics of `Ok` will put additional burden on optimizer because it will have to treat it same way as `Nil` in many but not all situations.

What is related to commonplace meaning of `Nil`, I wonder why people choose the second part of  its description over the first part? I mean _Absence of a value or a benign failure_ from https://docs.raku.org/type/Nil.

**Comment by vrurg** [comment]
> For returning nothing I usually use Empty, not Nil…

`Empty` is a positional (`Slip`) which doesn't make it reasonable for non-positional return types:

```
sub foo(--> Int) { Empty };
foo; # Type check failed for return value; expected Int but got Slip (Empty)
```

**Comment by tony-o** [comment]
What’s the issue with just changing the return types of those?

`+str` should return a not a number failure
`sleep` should return the number of seconds slept 
`EVAL` should return the last value like a block

having a type that’s the exact opposite of failure without more meaning can already be expressed using what’s there and a `!`.

**Comment by codesections** [comment]
> `EVAL` should return the last value like a block

Oops, that was a typo on my part (now fixed).  `EVAL($str)` _does_ return the last value; it's `EVAL($str, :check)` that returns `Nil`, to indicate that it checked the syntax without encountering any errors.

**Comment by vrurg** [comment]
I think @tony-o approach works best here. The concept of _meaningful nothing_ really sounds weird and over-complicates things. The need to instantiate `Ok` to get true out of it – why not just return `True` then? Perhaps there is a need to re-consider return values of some of the core subs and methods. But otherwise I'm personally still not convinced that `Ok` will bring more profit than harm.

As to the return values – `EVAL(:check)` returning `True` worth considering. Luckily, it's not specced yet.

And, BTW, `ok` is already [occupied by `Test`](https://docs.raku.org/type/Test#sub_ok) and we better don't touch it. ;)

**Comment by JJ** [comment]
I think this is similar to type None, used in Haskell and other languages,
right?

**Comment by niner** [comment]
The whole idea of an Ok or Success return value seems to be predicated on the 
assumption that it's actually necessary to check every single method or 
function call for success or failure. I don't see that assumption holding. If 
I call a function, I actually assume that it will do what it ought to do. If 
something really unexpected happens and it cannot perform its job, I expect it 
to throw an exception.

Now exceptions are sometimes a bit cumbersome to use and there are cases when 
exceptions are more likely than "it would be really odd if this happened but 
that's life I guess", e.g. IO failing due to permission errors or the like. 
For these cases we have Failure which can be checked for and defused by simple 
boolification as would happen in an if clause or boolean chain.

That should really cover all cases. Assume success, catch exceptions if 
there's a good way to handle them and use Failures when failure is an option 
but you still need to ensure that failures get handled somehow.

Because that's the whole point: to not silently ignore failing calls and have 
something else blow up later when it's much harder to discern what happened. 
An explicit "Success" return type would not help with that.

What else could it help with? Distinguishing between "did produce a value" and 
"could not produce a value" cases was brought up. But we already have a 
perfectly fine mechanism for that: defined vs. undefined. And conveniently Nil 
does become an undefined value of the appropriate type when assigned.

As to the examples that a Success type would help with:

I fear that EVAL(:check) is just not a well designed API anyway. A function or 
method should have one and only one purpose and complete exactly one task. 
With the :check argument, EVAL's purpose changes completely. Instead of 
executing some code and returning what the code evaluates to, it now checks 
the code for syntactical and grammatical correctness and returns nothing but 
throws an exception if the code is not correct.

I know that the implementation of both is quite similar, but that does not say 
anything about whether it's good API design to have them share a name. I 
completely understand the temptation to give the function additional 
functionality with such minor tweaks. Unfortunately that also leads to the 
second issue besides having a function called EVAL that doesn't actually 
always evaluate: we have a function that throws exceptions in normal use.

The fact that someone would want to actively check some code already implies 
that it's to be expected that not all code would pass this check (otherwise an 
exception during a normal EVAL would be enough to handle the...well 
exception). But exceptions are not meant to be thrown during normal 
processing. They are as the name states quite explicitly for the proverbial 
exception to the rule. As we'd expect EVAL(:check) to report errors in the 
checked code, the proper way to do so would be a return value. So it should be 
EVAL(:check --> Bool) instead of EVAL(:check --> Any).

Fixing the inappropriate use of exceptions as "return values" directly exposes 
the design issue: that an argument of the function changes the function's 
signature. That's confusing and clearly not good design. Of course in Raku we 
have a kind of escape hatch with multis, so we could have a multi EVAL($code, 
:$check! --> Bool) candidate, but that still smells and at that point we could 
as well give it a different name.

So long story short, tony-o is right that this is not really an example of the 
need for a Success type but for fixing a problem with the function itself.

I have to admit that I don't really understand the second example.
```raku
my %h is BagHash;
(try %h.add: +$str ) ?? #`(handle success) !! #`(handle failure);
```
This is already perfectly possible with Failures or exceptions and that seems 
to be what this example presumes (otherwise why have the try there?):
> rakudo -e '(try BagHash.new.add: +"f") ?? say "success" !! say "failure"'
failure

It seems to me that instead of adding yet more ways to distinguish between 
successful and failing operations, we would benefit of good documentation of 
the options we already have and where they are appropriate.

**Comment by lizmat** [comment]
FWIW, I take `--> Nil` to be the Raku equivalent of C's `void`.
